contracts/MemeFactory.sol

```diff
@@ -180,23 +182,68 @@ abstract contract MemeFactory {
         // Approve tokens for router
-        IERC20(nativeToken).approve(uniV2router, nativeTokenAmount);
-        IERC20(memeToken).approve(uniV2router, memeTokenAmount);
+        IERC20(nativeToken).approve(uniV3positionManager, nativeTokenAmount);
+        IERC20(memeToken).approve(uniV3positionManager, memeTokenAmount);
 
         // Add native token + meme token liquidity
-        (, , liquidity) = IUniswap(uniV2router).addLiquidity(
-            nativeToken,
-            memeToken,
-            nativeTokenAmount,
-            memeTokenAmount,
-            0, // Accept any amount of native token
-            0, // Accept any amount of meme token
-            address(this),
-            block.timestamp
-        );
+        INonfungiblePositionManager.MintParams memory params = INonfungiblePositionManager.MintParams(
+            token0: nativeToken,
+            token1: memeToken,
+            fee: feeTier,
+            tickLower: TickMath.MIN_TICK,
+            tickUpper: TickMath.MAX_TICK,
+            amount0Desired: nativeTokenAmount,
+            amount1Desired: memeTokenAmount,
+            amount0Min: 0, // Accept any amount of native token
+            amount1Min: 0, // Accept any amount of meme token
+            recipient: address(this),
+            deadline: block.timestamp
+        });
+
+        // Add native token + meme token liquidity
+        // (, , liquidity) = IUniswap(uniV2router).addLiquidity(
+        //     nativeToken,
+        //     memeToken,
+        //     nativeTokenAmount,
+        //     memeTokenAmount,
+        //     0, // Accept any amount of native token
+        //     0, // Accept any amount of meme token
+        //     address(this),
+        //     block.timestamp
+        // );
+        INonfungiblePositionManager(uniV3positionManager).mint(params);
+
+        // store token id for later
+
 
         // Get the pair address
-        pair = IUniswap(uniV2factory).getPair(nativeToken, memeToken);
+        // pair = IUniswap(uniV2factory).getPair(nativeToken, memeToken);
+    }
+
+    /// @dev Collects all accumulated LP fees in native token.
+    function collectFeesSingleSided() external {
+        for (uint256 i = 0; i < accounts.length; ++i) {
+            // FIX: loop over tokens - we should somehow allow for batching
+            // use _collectFees below
+        }
+    }
+
+    function _collectFees(tokenId) {
+        INonfungiblePositionManager.CollectParams memory params = INonfungiblePositionManager.CollectParams({
+            tokenId: tokenId,
+            recipient: address(this),
+            amount0Max: type(uint128).max,
+            amount1Max: 0
+        });
+
+        INonfungiblePositionManager(uniV3positionManager).collect(params);
+
+        uint256 nativeAmountForOLASBurn = // FIX: assign here the collected native token amount
+
+        uint256 adjustedNativeAmountForAscendance = _redemptionLogic(nativeAmountForOLASBurn);
```

Comment:
We reuse the same logic to ensure any and all native tokens collected go to redemption first